server/src/main/java/org/elasticsearch/script/CtxMap.java

```diff
@@ -52,26 +51,21 @@ protected CtxMap(Map<String, Object> source, Metadata metadata) {
         }
     }
 
-    protected Map<String, Object> wrapSource(Map<String, Object> source) {
-        Map<String, Object> wrapper = Maps.newHashMapWithExpectedSize(1);
-        wrapper.put(SOURCE, source);
```

Comment:
Why do we need more than replacing this wrapper with `return Map.of(SOURCE, source);`? Then the map is immutable, no keys can be written to the map that _source is in.

Reply:
it is valid to replace `_source` with another map representing the new source.

Reply:
I don't think using an immutable map precludes that requirement. If a script has `ctx._source = ['foo': 'bar']`, the put still happens on CtxMap. It can then completely replace the immutable source map in this special case, it doesn't need to delegate.

Reply:
Puts also happen in the EntrySet's Map.Entry which is available through entrySet() and entrySet().iterator().

Reply:
But that is still on the CtxMap right? Ultimately that is what controls everything under `ctx`, so *replacing* _source by doing a put, wherever that is coming from, should be controllable within CtxMap, without needing any special source map wrapper.

Reply:
I'm also not understanding the need for an additional indirection w/ `RestrictedMap` when this is already something `CtxMap` is specifically specialized for. Seems like `CtxMap` could easily handle the addition of the specialized key even if this needs to be done on a case by case per context with something like `IngestCtxMap`.

Reply:
That's wasn't too bad.  Any ideas on how we can ensure `source.put` is never called in `CtxMap`?